File: internal/metadataedit/picturetypes.go
package metadataedit

// PictureType is one kind of attached picture the editor manages. ID is the
// canonical TagLib pictureType string as stored in the files; FileBase is the
// filename base used both for folder art (<base>.jpg/png) and for the named
// assetstore entry.
type PictureType struct {
	ID       string
	Label    string
	FileBase string
}

// PictureTypes is the curated, ordered registry of picture types the editor
// offers. Mirrored in webui/src/lib/pictureTypes.ts — keep the two in sync.
var PictureTypes = []PictureType{
	{ID: "Front Cover", Label: "Front cover", FileBase: "cover"},
	{ID: "Back Cover", Label: "Back cover", FileBase: "back"},
	{ID: "Media", Label: "Media (disc)", FileBase: "disc"},
	{ID: "Leaflet Page", Label: "Booklet", FileBase: "booklet"},
	{ID: "Artist", Label: "Artist", FileBase: "artist"},
	{ID: "Band", Label: "Band", FileBase: "band"},
	{ID: "Illustration", Label: "Illustration", FileBase: "illustration"},
	{ID: "Other", Label: "Other", FileBase: "other"},
}

// PictureTypeByID looks a registry entry up by its canonical ID.
func PictureTypeByID(id string) (PictureType, bool) {
	for _, pt := range PictureTypes {
		if pt.ID == id {
			return pt, true
		}
	}
	return PictureType{}, false
}

File: webui/src/lib/pictureTypes.ts
import type { CoverCandidate, PictureSlot } from '@/types/metadata'

// The curated, ordered registry of attached-picture types the editor offers.
// IDs are the canonical TagLib pictureType strings as stored in the files.
// Mirrors internal/metadataedit/picturetypes.go — keep the two in sync.
export interface PictureTypeDef {
    id: string
    label: string
    // The matching Cover Art Archive image type, for sorting search results.
    caaType?: string
}

export const PICTURE_TYPES: readonly PictureTypeDef[] = [
    { id: 'Front Cover', label: 'Front cover', caaType: 'Front' },
    { id: 'Back Cover', label: 'Back cover', caaType: 'Back' },
    { id: 'Media', label: 'Media (disc)', caaType: 'Medium' },
    { id: 'Leaflet Page', label: 'Booklet', caaType: 'Booklet' },
    { id: 'Artist', label: 'Artist' },
    { id: 'Band', label: 'Band' },
    { id: 'Illustration', label: 'Illustration', caaType: 'Other' },
    { id: 'Other', label: 'Other', caaType: 'Other' }
] as const

export function pictureTypeLabel(id: string): string {
    return PICTURE_TYPES.find((t) => t.id === id)?.label ?? id
}

// The editor's slot display/preference order (embedded first) and labels. Both
// slots are on disk; this is NOT the app's serving precedence, which still
// prefers an uploaded cover in aether's store over the folder file.
export const PICTURE_SLOTS: readonly PictureSlot[] = ['embedded', 'folder'] as const

export const PICTURE_SLOT_LABELS: Record<PictureSlot, string> = {
    embedded: 'embedded in file',
    folder: 'album folder'
}

// candidateMatchesType reports whether a Cover Art Archive candidate depicts
// the requested picture type.
export function candidateMatchesType(candidate: CoverCandidate, typeId: string): boolean {
    const caa = PICTURE_TYPES.find((t) => t.id === typeId)?.caaType
    if (!caa) return false
    return candidate.types?.includes(caa) ?? false
}

// sortCandidatesForType orders CAA results for a picker scoped to one type:
// type matches first, then (for the front cover) isFront images, rest after,
// preserving the archive's order within each group.
export function sortCandidatesForType(
    candidates: CoverCandidate[],
    typeId: string
): CoverCandidate[] {
    const rank = (c: CoverCandidate): number => {
        if (candidateMatchesType(c, typeId)) return 0
        if (typeId === 'Front Cover' && c.isFront) return 1
        return 2
    }
    return [...candidates].sort((a, b) => rank(a) - rank(b))
}
